server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

```diff
@@ -0,0 +1,473 @@
+                dataStream.getName()
+            );
+
+            // we should reduce the number of shards
+            return new AutoShardingResult(
+                AutoShardingType.DECREASES_NUMBER_OF_SHARDS,
+                writeIndex.getNumberOfShards(),
+                Math.toIntExact(optimalReduceNumberOfShards),
+                remainingReduceShardsCooldown,
+                maxIndexLoadWithinCoolingPeriod
+            );
+        }
+
+        return new AutoShardingResult(
+            AutoShardingType.NO_CHANGE_REQUIRED,
+            writeIndex.getNumberOfShards(),
+            writeIndex.getNumberOfShards(),
+            TimeValue.ZERO,
+            writeIndexLoad
+        );
+    }
+
+    // Visible for testing
+    static long computeOptimalNumberOfShards(int minNumberWriteThreads, int maxNumberWriteThreads, double indexingLoad) {
+        return Math.max(
+            Math.min(roundUp(indexingLoad / ((double) minNumberWriteThreads / 2)), 3),
+            roundUp(indexingLoad / ((double) maxNumberWriteThreads / 2))
+        );
+    }
+
+    private static long roundUp(double value) {
+        return Math.round(Math.ceil(value));
+    }
+
+    // Visible for testing
+    static double getMaxIndexLoadWithinCoolingPeriod(
+        Metadata metadata,
+        DataStream dataStream,
+        double writeIndexLoad,
+        TimeValue coolingPeriod,
+        LongSupplier nowSupplier
+    ) {
+        // for reducing the number of shards we look at more than just the write index
+        List<IndexWriteLoad> writeLoadsWithinCoolingPeriod = DataStream.getIndicesWithinMaxAgeRange(
+            dataStream,
+            metadata::getIndexSafe,
+            coolingPeriod,
+            nowSupplier
+        )
+            .stream()
+            .filter(index -> index.equals(dataStream.getWriteIndex()) == false)
+            .map(metadata::index)
+            .filter(Objects::nonNull)
+            .map(IndexMetadata::getStats)
+            .filter(Objects::nonNull)
+            .map(IndexMetadataStats::writeLoad)
+            .filter(Objects::nonNull)
+            .toList();
+
+        // assume the current write index load is the highest observed and look back to find the actual maximum
+        double maxIndexLoadWithinCoolingPeriod = writeIndexLoad;
+        for (IndexWriteLoad writeLoad : writeLoadsWithinCoolingPeriod) {
+            double totalIndexLoad = 0;
+            for (int shardId = 0; shardId < writeLoad.numberOfShards(); shardId++) {
+                final OptionalDouble writeLoadForShard = writeLoad.getWriteLoadForShard(shardId);
+                if (writeLoadForShard.isPresent()) {
+                    totalIndexLoad += writeLoadForShard.getAsDouble();
+                }
+            }
+
+            if (totalIndexLoad > maxIndexLoadWithinCoolingPeriod) {
+                maxIndexLoadWithinCoolingPeriod = totalIndexLoad;
+            }
+        }
+        return maxIndexLoadWithinCoolingPeriod;
```

Comment:
I don't know whether this is any cleaner :) but I played some code golf:

```suggestion
        return Math.max(
            writeIndexLoad,
            writeLoadsWithinCoolingPeriod.stream()
                .mapToDouble(
                    load -> IntStream.range(0, load.numberOfShards())
                        .mapToObj(load::getWriteLoadForShard)
                        .mapToDouble(optLoad -> optLoad.orElse(0))
                        .sum()
                )
                .max()
                .orElse(0)
        );
```

Reply:
If you don't have a strong opinion here I'd rather keep a nested loop (as in PR) compare to streaming within a stream and mapping within a map :) 
Happy to apply the suggestion if you think it's more readable.